### PR TITLE
Load local file

### LINKED ISSUE #40: Allow a user to render chart using a local data file (for use within a single session)
Should be doable in new-ish browsers. Tutorials [here](https://www.html5rocks.com/en/tutorials/file/dndfiles/) and [here](https://developer.mozilla.org/en-US/docs/Web/API/File/Using_files_from_web_applications).

# Test Notes

Only officially supporting chrome for now, but this functionality should work in FF and Safari (and other modern browsers) as well. 
 
- New "Choose File" and "Load" controls under the "2. Choose a data set" heading should be shown on load. 
- Load button is disabled by default
- Info icon should display a note on hover. 
- Browse local files when clicking choose file
- When a .csv file is selected, the Load button should become active, and a note should appear saying file is ready to load
- Load button should stay disabled when any other file type is chosen, and a note should appear saying file can't be loaded
- After clicking load: 
  - The file should be added to the data set dropdown
  - A message saying the file was loaded should be shown
  - The file chooser should go back to "No file chosen" 
- Newly loaded files should work in renderers (assuming the data is properly formatted). Best way to test this is in web-codebook, which should renderer pretty much any csv less than a few megabytes.

### PR DESCRIPTION
**Pull request by jwildfire**
closes #40

### CODE REVIEW

File: src/cat/controls/initBootstrapConfig.js [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
+  var settingsHeading = cat.controls.environmentWrap
+    .append("h3")
+    .html("4. Environment ");
 
-  cat.controls.bootstrapButton = settingsSection
+  cat.controls.bootstrapButton = cat.controls.environmentWrap
     .append("button")
     .text("Load Bootstrap")
     .on("click", function() {
       loadBootstrap(cat);
```
Comment (samussiah) [inline src/cat/controls/initBootstrapConfig.js]: This is calling `loadRenderer()` but not clearing out the current chart.  Does the chart need to be re-rendered?

File: src/cat/renderChart.js [hunk 2]
```diff
@@ -7,12 +7,12 @@ export function renderChart(cat) {
   cat.settings.sync(cat);
   //render the new chart with the current settings
   var dataFile = cat.controls.dataFileSelect.node().value;
-  var dataFilePath = cat.config.dataURL + dataFile;
+  var dataObject = cat.config.dataFiles.find(f => f.label == dataFile);
```
Comment (samussiah) [inline src/cat/renderChart.js]: add this polyfill for find to support IE: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find#Polyfill

File: src/cat/controls/initBootstrapConfig.js (hunk 1 above)
Comment (samussiah) [inline src/cat/controls/initBootstrapConfig.js]: This is calling `loadRenderer()` but not clearing out the current chart.  Does the chart need to be re-rendered?

File: src/cat/controls/initFileLoad.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    .attr("class", "loadStatus")
+    .style("float", "right")
+    .text("Select a csv to load");
+
+  cat.controls.dataFileLoad = cat.controls.dataWrap
+    .append("input")
+    .attr("type", "file")
+    .attr("class", "file-load-input")
+    .on("change", function() {
+      if (this.value.slice(-4) == ".csv") {
```
Comment (samussiah) [inline src/cat/controls/initFileLoad.js]: Maybe `toLowerCase()` this just to cover all the bases.

File: src/cat/controls/initFileLoad.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        //shouldn't happen since button is disabled when no file is present, but ...
+        console.log("No file selected ...");
+        return false;
+      }
+
+      var fr = new FileReader();
+      fr.onload = function(e) {
+        //make sure the file is a csv
+
+        //make an object for the file
```
Comment (samussiah) [inline src/cat/controls/initFileLoad.js]: Very cool.

File: src/cat/controls/initFileLoad.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        var dataObject = {
+          label: files[0].name + " (loaded by user)",
```
Comment (samussiah) [inline src/cat/controls/initFileLoad.js]: Perhaps check the label against the current set of labels and add a _(1)_ or something so the user can upload two files with the same name, like when they're editing the same data file to test edge cases.

File: src/cat/setDefaults.js [hunk 6]
```diff
@@ -6,4 +6,11 @@ export function setDefaults(cat) {
   cat.config.dataURL = cat.config.dataURL || defaultSettings.dataURL;
   cat.config.dataFiles = cat.config.dataFiles || defaultSettings.dataFiles;
   cat.config.renderers = cat.config.renderers || defaultSettings.renderers;
+
+  cat.config.dataFiles = cat.config.dataFiles.map(function(df) {
+    return typeof df == "string"
+      ? { label: df, path: cat.config.dataURL, user_loaded: false }
+      : df;
+  });
+  console.log(cat.config.dataFiles);
```
Comment (samussiah) [inline src/cat/setDefaults.js]: stray clog

File: src/cat/controls/initBootstrapConfig.js (hunk 1 above)
Comment (jwildfire) [reply src/cat/controls/initBootstrapConfig.js]: Maybe ... but `initBoostrapConfig.js` goes away in favor of a more generalized approach in a downstream PR, so not going to worry about it ;)

File: src/cat/renderChart.js (hunk 2 above)
Comment (jwildfire) [reply src/cat/renderChart.js]: Fixed in `hotfixes` branch.

File: src/cat/controls/initFileLoad.js (hunk 3 above)
Comment (jwildfire) [reply src/cat/controls/initFileLoad.js]: Fixed in `hotfixes` branch.

File: src/cat/controls/initFileLoad.js (hunk 5 above)
Comment (jwildfire) [reply src/cat/controls/initFileLoad.js]: added timestamp to distinguish duplicates in `hotfixes` branch

File: src/cat/setDefaults.js (hunk 6 above)
Comment (jwildfire) [reply src/cat/setDefaults.js]: Cleaned up in downstream PR

**Comment by jwildfire** [comment]
Thanks for the nice review @samussiah. Fixed the review comments in a new `hotfixes` branch to avoid merge conflicts. Merging this to dev.
